### PR TITLE
[lldb] Cherry pick expression and dwim-print updates

### PR DESCRIPTION
**Pull request by kastiglione**
Imports the latest dwim-print related commits from llvm.org.

### CODE REVIEW

**Comment by kastiglione** [comment]
@swift-ci test

File: lldb/source/Commands/CommandObjectExpression.cpp [hunk 1]
```diff
@@ -144,8 +145,21 @@ Status CommandObjectExpression::CommandOptions::SetOptionValue(
...
+      suppress_persistent_result = !persist_result;
+    else
+      error.SetErrorStringWithFormat(
+          "could not convert \"%s\" to a boolean value.",
+          option_arg.str().c_str());
+    break;
+  }
+
+  // BEGIN SWIFT
+  case '\x02': {
```
Comment (kastiglione) [inline lldb/source/Commands/CommandObjectExpression.cpp]: @augusto2112 does this seem ok to you?

**Comment by kastiglione** [comment]
@swift-ci test

File: lldb/source/Commands/CommandObjectExpression.cpp (hunk 1 above)
Comment (augusto2112) [reply lldb/source/Commands/CommandObjectExpression.cpp]: Yes, this looks fine :)

**Comment by kastiglione** [comment]
@swift-ci test

**Comment by kastiglione** [comment]
@swift-ci test macOS Platform

**Comment by kastiglione** [comment]
The single test failure (TestSwiftRegex.py) doesn't appear to be related to this change.

```
AssertionError: Ran command:
"po regex"

Got output:

Expecting sub string: "Regex<(Substring, Substring, Substring, Substring)>" (was not found)
```

File: lldb/source/Commands/Options.td [hunk 2]
```diff
@@ -387,8 +387,13 @@ let Command = "expression" in {
...
     Desc<"Controls whether the expression can fall back to being JITted if it's "
     "not supported by the interpreter (defaults to true).">;
+  def persistent_result : Option<"persistent-result", "\\x01">, Groups<[1,2]>,
+    Arg<"Boolean">,
+    Desc<"Persist expression result in a variable for subsequent use. "
+    "Expression results will be labeled with $-prefixed variables, e.g. $0, "
+    "$1, etc. Defaults to true.">;
   // BEGIN SWIFT
-  def bind_generic_types : Option<"bind-generic-types", "\\x01">,
+  def bind_generic_types : Option<"bind-generic-types", "\\x02">,
```
Comment (adrian-prantl) [inline lldb/source/Commands/Options.td]: Could we use a much higher number for this, so it's less likely to break in the future?

Comment (kastiglione) [reply lldb/source/Commands/Options.td]: Good idea.

**Comment by kastiglione** [comment]
Testing in https://github.com/apple/llvm-project/pull/6309

File: lldb/source/Commands/Options.td (hunk 2 above)
Comment (kastiglione) [reply lldb/source/Commands/Options.td]: Switched to `\x31`, the highest control code (Unit Separator) and last char value prior to the first printable character (space character).

Comment (kastiglione) [reply lldb/source/Commands/Options.td]: cc @augusto2112

**Comment by kastiglione** [comment]
@swift-ci test

File: lldb/source/Commands/Options.td (hunk 2 above)
Comment (augusto2112) [reply lldb/source/Commands/Options.td]: Yes I think that's fine!

**Comment by kastiglione** [comment]
Seems the regex test failure is caused by these changes. Will investigate

**Comment by kastiglione** [comment]
@swift-ci test macOS platform

**Comment by kastiglione** [comment]
@swift-ci test

**Comment by kastiglione** [comment]
From https://github.com/apple/llvm-project/pull/6347, the problem commit is "[lldb] Suppress persistent result when running po".

**Comment by kastiglione** [comment]
@swift-ci test

**Comment by kastiglione** [comment]
@swift-ci test Linux platform

**Comment by kastiglione** [comment]
@swift-ci test

**Comment by kastiglione** [comment]
@swift-ci test macOS Platform

**Comment by kastiglione** [comment]
@swift-ci test

**Pull request merged**